algebird-core/src/main/scala/com/twitter/algebird/CorrelationGroup.scala

```diff
@@ -0,0 +1,165 @@
+
+  /**
+   * Given two streams of doubles (weightN, an) and (weightK, ak) of form (weighted count,
+   * mean), calculates the mean of the combined stream.
+   *
+   * Uses a more stable online algorithm which should be suitable for
+   * large numbers of records similar to:
+   * http://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Parallel_algorithm
+   *
+   * This differs from the implementation in MomentsGroup.scala only in that here, the counts are weighted, and are
+   * thus doubles instead of longs
+   */
+  def getCombinedMean(weightN: Double, an: Double, weightK: Double, ak: Double): Double =
+    if (weightN < weightK) getCombinedMean(weightK, ak, weightN, an)
+    else
+      (weightN + weightK) match {
+        case 0L                              => 0.0
+        case newCount if newCount == weightN => an
+        case newCount =>
+          val scaling = weightK.toDouble / newCount
+          // a_n + (a_k - a_n)*(k/(n+k)) is only stable if n is not approximately k
+          if (scaling < STABILITY_CONSTANT) (an + (ak - an) * scaling)
+          else (weightN * an + weightK * ak) / newCount
+      }
+
+}
+
+/**
+ * A class to calculate covariance and the first two central moments of a sequence of pairs of Doubles, from which the
+ * pearson correlation coeifficient can be calculated.
+ *
+ * m{i}Left denotes the ith central moment of the first projection of the pair.
+ * m{i}Right denotes the ith central moment of the second projection of the pair.
+ * c2 the covariance equivalent of the second central moment, i.e. c2 = Sum_(l,r) (l - m1Left)*(r - m1Right).
+ *
+ */
+case class Correlation(
+    c2: Double,
+    m2Left: Double,
+    m2Right: Double,
+    m1Left: Double,
+    m1Right: Double,
+    m0: Double
+) {
+  def totalWeight: Double = m0
+
+  def meanLeft: Double = m1Left
+
+  def meanRight: Double = m1Right
+
+  // variance, stddev, and covariance are for the population, not a sample
+
+  def varianceLeft: Double = m2Left / m0
+
+  def varianceRight: Double = m2Right / m0
+
+  def stddevLeft: Double = Math.sqrt(varianceLeft)
+
+  def stddevRight: Double = Math.sqrt(varianceRight)
+
+  def covariance: Double = c2 / totalWeight
+
+  /**
+   *
+   * @return Pearson's correlation coefficient
+   */
+  def correlation: Double =
+    // correlation is defined as: covariance / (varianceLeft * varianceRight)
+    // however, dividing by "count" cancels out, and leaves us with the following formula, which relies on fewer
+    // divisions
+    c2 / (Math.sqrt(m2Left * m2Right))
+
+  /**
+   * Assume this instance of Correlation came from summing together Correlation.apply((x_i, y_i)) for i in 1...n.
+   * @return (m, b) where y = mx + b is the line with the least squares fit of the points (x_i, y_i).
+   *         See, e.g. https://mathworld.wolfram.com/LeastSquaresFitting.html.
+   */
+  def linearLeastSquares: (Double, Double) = {
+    val m = c2 / m2Left
+    val b = meanRight - m * meanLeft
+    (m, b)
+  }
+
```

Comment:
now that we have m0 as a Double, can we add `def scale(z: Double): Correlation` which scales the entire correlation (which is used for making exponential decays)?

I think the mean isn't touched but the totalWeight and the second moments are changed.

The laws we want: `x.scale(0.0) == Monoid.zero`, `x.scale(a).scale(b) == x.scale(a * b)`, `x.scale(1.0) == x`. I *think* those laws are enough...

(obviously with the correct approximately equals situation).

Reply:
The linked paper seems to agree with you, but those laws are also satisfied by either a) making `scale` a no-op or b) only scaling `m0` (which it took me a while to convince myself was the wrong thing). Should there also be an invariant about the average-like quantities (`meanX`, `meanY`, `varianceX`, `varianceY`) being unchanged after scaling? That would at least rule out option b.

Reply:
I think no-op scale fails `x.scale(0.0) == zero`. So, I don't think it can be a no-op.

Reply:
Ok, that works. Just pushed.